modules/node-shared/src/test/scala/org/tessellation/node/shared/domain/seedlist/SnapshotOrdinalTimelineSuite.scala

```diff
@@ -0,0 +1,98 @@
+package org.tessellation.node.shared.domain.seedlist
```

Comment:
@cngo-github (cc: @sdavidp)

1. You need to add `@derive(eqv)` to be able to use `expect.eql`.
2. I see that you decoupled the decoders of single and multiple ranges but looks like you reuse the same error types. Please do the following and you should spot the issue:

```scala
@derive(show, eqv)                                                                                                                                                                                                   
sealed trait RangeError extends NoStackTrace                                                                                                                                                                         
                                                                                                                                                                                                                     
case class InvalidOrdinal(ordinal: String) extends RangeError {                                                                                                                                                      
  override def getMessage = s"Invalid ordinal: $ordinal"                                                                                                                                                             
}                                                                                                                                                                                                                    
                                                                                                                                                                                                                     
case class InvalidRangeFormat(range: List[String]) extends RangeError {                                                                                                                                              
  override def getMessage = s"Invalid range format: ${range.mkString("-")}"                                                                                                                                          
}                                                                                                                                                                                                                    
                                                                                                                                                                                                                     
@derive(show)                                                                                                                                                                                                        
@derive(show, eqv)                                                                                                                                                                                                   
case class Range(start: NonNegLong, end: NonNegLong)                                                                                                                                                                 
                                                                                                                                                                                                                     
object Range {                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                                                                           
  type ValidationResult[A] = ValidatedNel[RangeError, A]  
```
and equivalent for `RangesError` in `ranges.scala` file.

You would need to change `.traverse(Range.decode)` to `.traverse(r => Range.decode(r).leftMap(_.map(InvalidRange)))` and then you should have it decoupled correctly.

3. You check only happy-paths in the unit tests. IMO one of main goals of unit tests is to catch errors and check if non happy paths are correctly handled. Please see the following tests, some of them are left to be implemented, as I had no more time to focus on this. I've implemented one or two error cases, so the rest would be mostly copy-paste.

4. Unit tests are bloated and hard to read, hence hard to maintain. Please see the revised tests:
```scala
object RangeSuite extends FunSuite with Checkers {  
 implicit class DecodeOps(input: String) {                                                               
    def asRange: Option[Range] = Range.decode(input).toOption                                             
  }  

  test("decodes terminated range") {                                                                      
    val result = "10-15".asRange                                                                          
    val expected = Range(10L, 15L).some                                                                   
                                           
    expect.eql(expected, result)                                                                          
  }    

  test("ignores leading and trailing spaces") {                                                           
    val result = "  20  -  30  ".asRange                                                                  
    val expected = Range(20L, 30L).some  

    expect.eql(expected, result)                                                                          
  }                                                                                                       

  test("decodes unterminanted range and assumes the end as Long.MaxValue") {                              
    val result = "11".asRange                                                                             
    val expected = Range(11L, Long.MaxValue).some 

    expect.eql(expected, result)                                                                          
  }    

  test("decodes unterminanted range with delimiter") {                                                    
    val result = "12-".asRange                                                                            
    val expected = Range(12L, Long.MaxValue).some  

    expect.eql(expected, result)                                                                          
  }                                                                                                       
                                                                                                          
  test("decodes descending range") {                                                                      
    val result = "99-60".asRange                                                                          
    val expected = Range(99L, 60L).some                                                                   
                                                                
    expect.eql(expected, result)                                                                          
  }    

  test("returns InvalidOrdinal for non-long entry") {                                                     
    val result = "a-6".asRange                                                                            
                                                                                                          
    ???                                                                                                   
  }  

  test("returns InvalidOrdinal for negative long entry") {                                                
    val result = "-34".asRange // TODO: not sure                                                          

    ???                                                                                                   
  }   

  test("returns InvalidRangeFormat for more than one or two long entries") {                              
    val result = "1-2-3".asRange              
                                                                
    ???                                                                                                   
  }  
}
```
and
```scala
object RangesSuite extends FunSuite with Checkers {                                                       
                                                                                                          
  implicit class DecodeOps(input: String) {                                                               
    def asRanges: Option[NonEmptyList[Range]] = ranges.decode(input).toOption                             
  }                                                                                                       
                                                                                                          
  implicit class ExpectedOps(ranges: List[(Long, Long)]) {                                                
    def asExpected: Option[NonEmptyList[Range]] =                                                         
      NonEmptyList                                                                                        
        .fromListUnsafe(ranges)                                                                           
        .map {                                                                                            
          case (start, end) => Range(NonNegLong.unsafeFrom(start), NonNegLong.unsafeFrom(end))                                                                                                                       
        }   
        .some     

  test("decodes multiple ranges") {                                                                       
    val result = "10-15;30-60".asRanges                                                                   
    val expected = List((10L, 15L), (30L, 60L)).asExpected                                                
                                                                                                          
    expect.eql(expected, result)                                                                          
  }   

  test("decodes ranges with trailing delimiter") {                                                        
    val result = "99-100;150-300;".asRanges                                                               
    val expected = List((99L, 100L), (150L, 300L)).asExpected 

    expect.eql(expected, result)                                                                          
  }  

  test("merges overlapping ranges") {                                                                     
    val result = "0-20;20-50;60-69;62-65;70-79".asRanges                                                  
    val expected = List((0L, 50L), (60L, 79L)).asExpected                                                 
                                                                                                          
    expect.eql(expected, result)                                                                          
  }  

  test("merges adjacent ranges inclusively") {                                                            
    val result = "0-19;20-34;35-60".asRanges                                                              
    val expected = List((0L, 60L)).asExpected                                                             
                                                                                                          
    expect.eql(expected, result)                                                                          
  }                                                                                                       
                                                                                                          
  test("excludes descending ranges") {                                                                    
    val result = "30-60;70-40;80-100".asRanges                                                            
    val expected = List((30L, 60L), (80L, 100L)).asExpected                                               
                                                                                                          
    expect.eql(expected, result)                                                                          
  }  

  test("returns InvalidInput if all ranges are descending") {                                             
    val input = "50-22;199-67;300-201"                                                                    
                                                              
    val result = ranges.decode(input).toEither                                                            
                                                                                                          
    result match {                                                                                        
      case Left(errors) => expect(errors.forall(_.isInstanceOf[InvalidInput]))                            
      case _            => failure("should return an error")                                              
    }                                                                                                     
  }  

  test("returns InvalidRange for invalid single range entry") {                                           
    val input = "12-34;36-7 8"                                                                            
                                                                                                          
    ???                                                                                                   
  }                                                                                                       
                                                                                                          
  test("returns InvalidRange if range is empty") {                                                        
    val result = ranges.decode("").toEither                                                               
                                                                                                          
    result match {                                                                                        
      case Left(errors) => expect(errors.forall(_.isInstanceOf[InvalidRange]))                            
      case _            => failure("should return an error")                                              
    }                                                                                                     
                                                                                                          
  }                                                                                                       
                                                                                                          
  test("filter out invalid ranges") {                                                                     
    val result = "0-20;20-50;60-69;62-65;70-79;89-80".asRanges                                            
    val expected = List((0L, 50L), (60L, 79L)).asExpected                                                 
                                                                                                          
    expect.same(expected, result)                                                                         
  }                                                                                                       

}  
```

5. `SnapshotOrdinalTimelineSuite` can be removed at all. Tests for `SnapshotOrdinalTimeline`  are in fact tests for `ranges`, as there is no additional logic in `SnapshotOrdinalTimeline`.

Reply:
One issue I made is that `.forall` for an empty list would return `true`. Probably it should be `.exists` instead

Reply:
1. Added `@derive(eqv, show)` to the errors and case classes as appropriate.
2. Re-worked the validation types into

```scala
  private type RangeValidationResult[A] = ValidatedNel[RangeError, A]
```

and

```scala
  private type RangesValidationResult[A] = ValidatedNel[RangesError, A]
```

Also, re-worked the errors as appropriate.

3. Re-worked the tests to cover more than just the happy path.
4. Re-worked the tests as shown.
5. Deleted `SnapshotOrdinalTimelineSuite`.